allennlp/models/semantic_parsing/atis/atis_semantic_parser.py

```diff
@@ -0,0 +1,592 @@
+                # linking scores are stored.
+                flattened_entity_index = batch_index * num_entities + entity_index
+                entity_types[flattened_entity_index] = entity_type
+            padded = pad_sequence_to_length(types, num_entities, lambda: 0)
+            batch_types.append(padded)
+
+        return tensor.new_tensor(batch_types, dtype=torch.long), entity_types
+
+    @staticmethod
+    def _action_history_match(predicted: List[int], targets: torch.LongTensor) -> int:
+        # TODO(mattg): this could probably be moved into a FullSequenceMatch metric, or something.
+        # Check if target is big enough to cover prediction (including start/end symbols)
+        if len(predicted) > targets.size(1):
+            return 0
+        predicted_tensor = targets.new_tensor(predicted)
+        targets_trimmed = targets[:, :len(predicted)]
+        # Return 1 if the predicted sequence is anywhere in the list of targets.
+        return torch.max(torch.min(targets_trimmed.eq(predicted_tensor), dim=1)[0]).item()
+
+    @staticmethod
+    def _postprocess_query_sqlite(query: str):
+        # The dialect of SQL that SQLite takes is not exactly the same as the labeled data.
+        # We strip off the parentheses that surround the entire query here.
+        query = query.strip()
+        if query.startswith('('):
+            return query[1:query.rfind(')')] + ';'
+        return query
+
+    @staticmethod
+    def action_sequence_to_sql(action_sequences: List[str]) -> str:
+        # Convert an action sequence like ['statement -> [query, ";"]', ...] to the
+        # SQL string.
+        query = []
+        for action in action_sequences:
+            nonterminal, right_hand_side = action.split(' -> ')
+            right_hand_side_tokens = right_hand_side[1:-1].split(', ')
+            if nonterminal == 'statement':
+                query.extend(right_hand_side_tokens)
+            else:
+                for query_index, token in reversed(list(enumerate(query))):
+                    if token == nonterminal:
+                        query = query[:query_index] + \
+                                right_hand_side_tokens + \
+                                query[query_index + 1:]
+                        break
+        return ' '.join([token.strip('"') for token in query])
+
+    @staticmethod
+    def is_nonterminal(token: str):
+        if token[0] == '"' and token[-1] == '"':
+            return False
+        return True
+
+    def _sql_result_match(self, predicted_query: str, sql_query_labels: List[str]) -> int:
+        postprocessed_predicted_query = self._postprocess_query_sqlite(predicted_query)
+
+        try:
+            self._cursor.execute(postprocessed_predicted_query)
+            predicted_rows = self._cursor.fetchall()
+        except sqlite3.Error as error:
+            logger.debug(f'Error executing predicted: {error}')
+            exit(0)
```

Comment:
Should you just `raise` here?  Calling `exit()` is generally not a good idea, unless you're writing a shell script.  Looks like the above line should also be `logger.error()`, right?

Reply:
Oh!  This is used in the subprocess.  This definitely should have a docstring explaining a bit more about what's going on.  And in that case, yes, `logger.debug()` is the right thing here.

Reply:
Ah yes, I'll add a comment -- we only ever call this in the subprocess.